torchtune/modules/loss/kd_losses.py

```diff
@@ -58,6 +58,105 @@ def forward(
             return torch.tensor(0.0, device=x.device)
         return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)
 
+class ReverseKLLoss(torch.nn.Module):
+    """
+    The Kullback-Leibler divergence loss for valid indexes.
+    Implementation of https://github.com/jongwooko/distillm/blob/17c0f98bc263b1861a02d5df578c84aea652ee65/distillm/losses.py
+
+    Args:
+        ignore_index (int):  Specifies a target value that is ignored and does not contribute to the input gradient.
+            The loss is divided over non-ignored targets.
+            Default: -100.
+    """
+
+    def __init__(self, ignore_index: int = -100):
+        super().__init__()
+        self.ignore_index = ignore_index
+
+    def forward(
+        self,
+        student_logits: torch.Tensor,
+        teacher_logits: torch.Tensor,
+        labels: torch.Tensor,
+        normalize: bool = True,
+    ) -> torch.Tensor:
+        """
+        Args:
+            student_logits (torch.Tensor): logits from student model of shape
+                (batch_size*num_tokens, vocab_size).
+            teacher_logits (torch.Tensor): logits from teacher model of shape
+                (batch_size*num_tokens, vocab_size).
+            labels (torch.Tensor): Ground truth labels of shape
+                (batch_size, vocab_size).
+            normalize (bool): Whether to normalize the loss by the number of unmasked elements.
+
+        Returns:
+            torch.Tensor: KL divergence loss of shape (1,).
+        """
+
+        student_prob = F.softmax(student_logits, dim=-1, dtype=torch.float32)
+
+        student_logprob = F.log_softmax(student_logits, dim=-1, dtype=torch.float32)
+        teacher_logprob = F.log_softmax(teacher_logits, dim=-1, dtype=torch.float32)
+
+        inf_mask = torch.isinf(student_logits) | torch.isinf(teacher_logits)
+
+        prod_probs = torch.masked_fill(student_prob * teacher_logprob, inf_mask, 0)
+        prod_probs -= torch.masked_fill(student_prob * student_logprob, inf_mask, 0)
+        x = torch.sum(prod_probs, dim=-1).view(-1)
+
+        mask = (labels != self.ignore_index).int()
+        if not normalize:
+            return -torch.sum(x * mask.view(-1), dim=0)
+        if torch.sum(mask.view(-1), dim=0) == 0:
+            return torch.tensor(0.0, device=x.device)
+        return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)
+
+class SymmetricKLLoss(torch.nn.Module):
+    """
+    The Symmetric Kullback-Leibler divergence loss for valid indexes.
+    Implementation of https://github.com/jongwooko/distillm/blob/17c0f98bc263b1861a02d5df578c84aea652ee65/distillm/losses.py
+
+    Args:
+        sym_kd_ratio (float): Ratio of symmetric KL divergence loss.
```

Comment:
nit: let's be a bit more explicit, e.g. "When set to 1 this loss reduces to forward KL divergence, when set to 0 this loss reduces to reverse kl divergence". Also separately it'd be good to do a value check that `0 <= sym_kd_ratio <= 1` on init.